### PR TITLE
Updated C# guidelines

### PR DESCRIPTION
**Pull request by edmondsb**
@MarkMichaelis  @adamskt  @Keboo 
Updated the csharp.md file to have the new guidelines, and added links to different sections headers. The other file updated was an image file that I used to see what examples might look like.

### CODE REVIEW

File: docs/coding/csharp.md [hunk 1]
```diff
@@ -1,141 +1,363 @@
...
+  - [Files](#files)
+  - [Whitespace](#whitespace)
+
+- [Coding](#coding)
+  - [Arrays](#arrays)
+  - [Assemblies](#assemblies)
+  - [Branches](#branches)
+  - [Classes](#classes)
+  - [Conversions](#conversions)
+  - [Dipsose()](#dipsose)
```
Comment (adamskt) [inline docs/coding/csharp.md:24]: Typo?

File: docs/coding/csharp.md [hunk 2]
```diff
@@ -1,141 +1,363 @@
...
+- :heavy_check_mark: DO: capitalize only the first character in acronyms with three or more characters, except for the first word of a camelCased identifier.
+- :x: DO NOT: capitalize any of the characters in acronyms at the beginning of a camelCased identifier.
+- :heavy_check_mark: DO: use PascalCasing for all class names.
+- :heavy_check_mark: DO: use camelCasing for local variable names.
+- :heavy_check_mark: DO: use uppercase literal suffixes (e.g., 1.618033988749895M).
+- :heavy_check_mark: DO: use camelCasing for variable declarations using tuple syntax. 
+- :heavy_check_mark: CONSIDER: using PascalCasing for all tuple item names.
+- :heavy_check_mark: DO: use PascalCasing for namespace names.
+- :heavy_check_mark: DO: use camelCasing for parameter names.
+- :heavy_check_mark: CONSIDER: using the same casing on a propertyâ€™s backing field as that used in the property, distinguishing the backing field with an â€œ_â€ prefix. Do not, however, use two underscores; identifiers beginning with two underscores are reserved for the use of t.
```
Comment (adamskt) [inline docs/coding/csharp.md:62]: Can we filter out the "smart" quotes that are coming from the CSV file? `with an â€œ_â€ prefix`

File: docs/coding/csharp.md (hunk 1 above)
Comment (edmondsb) [reply docs/coding/csharp.md:24]: yep

File: docs/coding/csharp.md [hunk 3]
```diff
@@ -1,141 +1,363 @@
...
+- :heavy_check_mark: DO: capitalize only the first character in acronyms with three or more characters, except for the first word of a camelCased identifier.
+- :x: DO NOT: capitalize any of the characters in acronyms at the beginning of a camelCased identifier.
+- :heavy_check_mark: DO: use PascalCasing for all class names.
+- :heavy_check_mark: DO: use camelCasing for local variable names.
+- :heavy_check_mark: DO: use uppercase literal suffixes (e.g., 1.618033988749895M).
+- :heavy_check_mark: DO: use camelCasing for variable declarations using tuple syntax. 
+- :grey_question: CONSIDER: using PascalCasing for all tuple item names.
+- :heavy_check_mark: DO: use PascalCasing for namespace names.
+- :heavy_check_mark: DO: use camelCasing for parameter names.
+- :grey_question: CONSIDER: using the same casing on a propertyâ€™s backing field as that used in the property, distinguishing the backing field with an â€œ_â€ prefix. Do not, however, use two underscores; identifiers beginning with two underscores are reserved for the use of t.
```
Comment (Keboo) [inline docs/coding/csharp.md:62]:
```suggestion
- :grey_question: CONSIDER: using the same casing on a property's backing field as that used in the property, distinguishing the backing field with an "_" prefix. Do not, however, use two underscores; identifiers beginning with two underscores are reserved for the use of t.
```

File: docs/coding/csharp.md [hunk 4]
```diff
@@ -1,141 +1,363 @@
...
+- :no_entry: AVOID: naming fields with camelCase.
+- :heavy_check_mark: DO: name properties with PascalCase.
+- :heavy_check_mark: DO: use PascalCasing and an â€œIâ€ prefix for interface names.
```
Comment (Keboo) [inline docs/coding/csharp.md:65]:
```suggestion
- :heavy_check_mark: DO: use PascalCasing and an "I" prefix for interface names.
```

File: docs/coding/csharp.md [hunk 5]
```diff
@@ -1,141 +1,363 @@
...
-- :heavy_check_mark: DO: Use unit test names that identify what is being tested
-- :heavy_check_mark: CONSIDER: Naming test methods with pattern <MethodName>_<StateUnderTest>_<ExpectedBehavior>
+- :heavy_check_mark: DO: favor clarity over brevity when naming identifiers.
+- :x: DO NOT: use Hungarian notation (that is, do not encode the type of a variable in its name).
+- :heavy_check_mark: DO: name classes with nouns or noun phrases.
+- :heavy_check_mark: DO: use the C# keyword rather than the BCL name when specifying a data type (e.g., string rather than String).
+- :x: DO NOT: use a constant for any value that can possibly change over time. The value of pi and the number of protons in an atom of gold are constants; the price of gold, the name of your company, and the version number of your program can change.
+- :heavy_check_mark: DO: name properties using a noun, noun phrase, or adjective.
+- :heavy_check_mark: DO: favor prefixing Boolean properties with “Is,” “Can,” or “Has,” when that practice adds value.
```
Comment (Keboo) [inline docs/coding/csharp.md:87]:
```suggestion
- :heavy_check_mark: DO: favor prefixing Boolean properties with "Is," "Can," or "Has," when that practice adds value.
```

File: docs/coding/csharp.md [hunk 6]
```diff
@@ -1,141 +1,363 @@
...
+- :x: DO NOT: include “sentinel” values (such as a value called Maximum); such values can be confusing to the user.
 
 ### Miscellaneous
-- :x: DO NOT: Use underscores to break up words in an identifier. Test methods, UI generated event handlers and references to Resource files are the exception.
-- :heavy_check_mark: DO: Use American-style spelling.
+- :heavy_check_mark: DO: treat parameter names as part of the API, and avoid changing the names if version compatibility between APIs is important.
+- :heavy_check_mark: DO: name the source file with the name of the public type it contains.
+- :grey_question: CONSIDER: giving a property the same name as its type.
+- :heavy_check_mark: DO: use the same name for constructor parameters (camelCase) and properties (PascalCase) if the constructor parameters are used to simply set the property.
+- :heavy_check_mark: DO: name an exception classes with the “Exception” suffix.
```
Comment (Keboo) [inline docs/coding/csharp.md:98]:
```suggestion
- :heavy_check_mark: DO: name an exception classes with the "Exception" suffix.
```

File: docs/coding/csharp.md [hunk 7]
```diff
@@ -1,141 +1,363 @@
...
+- :no_entry: AVOID: calling Dispose() on owned objects that have a finalizer. Instead, rely on the finalization queue to clean up the instance.
+- :heavy_check_mark: DO: invoke a base class’s Dispose() method when overriding Dispose().
+- :grey_question: CONSIDER: ensuring that an object becomes unusable after Dispose() is called. After an object has been disposed, methods other than Dispose() (which could potentially be called multiple times) should throw an ObjectDisposedException.
+- :heavy_check_mark: DO: implement IDisposable on types that own disposable fields (or properties) and dispose of those instances.
+- :heavy_check_mark: DO: invoke a base class’s Dispose() method from the Dispose(bool disposing) method if one exists.
+
+### Enums
+- :grey_question: CONSIDER: adding new members to existing enums, but keep in mind the compatibility risk.
+- :no_entry: AVOID: creating enums that represent an “incomplete” set of values, such as product version numbers.
+- :no_entry: AVOID: creating “reserved for future use” values in an enum.
```
Comment (Keboo) [inline docs/coding/csharp.md:168]:
```suggestion
- :no_entry: AVOID: creating "reserved for future use" values in an enum.
```

File: docs/coding/csharp.md [hunk 8]
```diff
@@ -1,141 +1,363 @@
...
+- :heavy_check_mark: DO: keep Dispose() simple, focusing on the resource cleanup required by finalization.
+- :no_entry: AVOID: calling Dispose() on owned objects that have a finalizer. Instead, rely on the finalization queue to clean up the instance.
+- :heavy_check_mark: DO: invoke a base class’s Dispose() method when overriding Dispose().
+- :grey_question: CONSIDER: ensuring that an object becomes unusable after Dispose() is called. After an object has been disposed, methods other than Dispose() (which could potentially be called multiple times) should throw an ObjectDisposedException.
+- :heavy_check_mark: DO: implement IDisposable on types that own disposable fields (or properties) and dispose of those instances.
+- :heavy_check_mark: DO: invoke a base class’s Dispose() method from the Dispose(bool disposing) method if one exists.
+
+### Enums
+- :grey_question: CONSIDER: adding new members to existing enums, but keep in mind the compatibility risk.
+- :no_entry: AVOID: creating enums that represent an “incomplete” set of values, such as product version numbers.
```
Comment (Keboo) [inline docs/coding/csharp.md:167]:
```suggestion
- :no_entry: AVOID: creating enums that represent an "incomplete" set of values, such as product version numbers.
```

File: docs/coding/csharp.md [hunk 9]
```diff
@@ -1,141 +1,363 @@
...
+- :no_entry: AVOID: creating “reserved for future use” values in an enum.
+- :no_entry: AVOID: enums that contain a single value.
+- :heavy_check_mark: DO: provide a value of 0 (none) for simple enums, knowing that 0 will be the default value when no explicit initialization is provided.
+- :heavy_check_mark: DO: use the FlagsAttribute to mark enums that contain flags.
+- :heavy_check_mark: DO: provide a None value equal to 0 for all flag enums.
+- :no_entry: AVOID: creating flag enums where the zero value has a meaning other than “no flags are set.”
```
Comment (Keboo) [inline docs/coding/csharp.md:173]:
```suggestion
- :no_entry: AVOID: creating flag enums where the zero value has a meaning other than "no flags are set."
```

File: docs/coding/csharp.md [hunk 10]
```diff
@@ -1,141 +1,363 @@
...
+- :no_entry: AVOID: catching and logging an exception before rethrowing it. Instead, allow the exception to escape until it can be handled appropriately.
+- :heavy_check_mark: DO: prefer using an empty throw when catching and rethrowing an exception so as to preserve the call stack.
+- :heavy_check_mark: DO: report execution failures by throwing exceptions rather than returning error codes.
+- :x: DO NOT: have public members that return exceptions as return values or an out parameter. Throw exceptions to indicate errors; do not use them as return values to indicate errors.
+- :x: DO NOT: use exceptions for handling normal, expected conditions; use them for exceptional, unexpected conditions.
+- :no_entry: AVOID: throwing exceptions from property getters.
+- :heavy_check_mark: DO: preserve the original property value if the property throws an exception.
+- :heavy_check_mark: DO: use nameof(value) (which resolves to “value”) for the paramName argument when creating ArgumentException() or ArgumentNullException() type exceptions (value" is the implicit name of the parameter on property setters)."
```
Comment (Keboo) [inline docs/coding/csharp.md:197]:
```suggestion
- :heavy_check_mark: DO: use nameof(value) (which resolves to "value") for the paramName argument when creating ArgumentException() or ArgumentNullException() type exceptions (value" is the implicit name of the parameter on property setters)."
```

**Review by Keboo — APPROVED** [review]

**Pull request merged**